File: Backend/renderer/src/main/java/net/chrisparton/sparkled/renderer/effect/FlashEffectRenderer.java
package net.chrisparton.sparkled.renderer.effect;

import net.chrisparton.sparkled.renderer.context.RenderContext;
import net.chrisparton.sparkled.renderer.util.FillUtils;

public class FlashEffectRenderer extends EffectRenderer {

    @Override
    public void render(RenderContext ctx) {
        float alpha = getAlpha(ctx.getProgress());

        for (int i = 0; i < ctx.getChannel().getLedCount(); i++) {
            FillUtils.fill(ctx, ctx.getFrame().getLed(i), alpha);
        }
    }

    /**
     * @param progress The 0 > 1 progress
     * @return The progress, transformed into 0 > 1 > 0
     */
    private float getAlpha(float progress) {
        float alpha = progress * 2;
        if (progress >= .5) {
            alpha = 1 - (progress - .5f) * 2;
        }

        // Remove floating point inaccuracies to ensure a symmetrical flash.
        return Math.round(alpha * 10000f) / 10000f;
    }
}

File: renderer/src/main/kotlin/io/sparkled/renderer/effect/FlashEffectRenderer.kt
package io.sparkled.renderer.effect

import io.sparkled.renderer.context.RenderContext
import io.sparkled.renderer.util.FillUtils

class FlashEffectRenderer : EffectRenderer() {

    override fun render(ctx: RenderContext) {
        val alpha = getAlpha(ctx.progress)

        for (i in 0 until ctx.channel.ledCount) {
            FillUtils.fill(ctx, i, alpha)
        }
    }

    /**
     * @param progress The normalised progress.
     *
     * @return The progress, transformed into 0 > 1 > 0.
     */
    private fun getAlpha(progress: Float): Float {
        val alpha = 2 * if (progress < .5) progress else 1 - progress

        // Remove floating point inaccuracies to ensure a symmetrical flash.
        return Math.round(alpha * 10000f) / 10000f
    }
}
